pyfar/dsp/dsp.py

```diff
@@ -2669,3 +2669,173 @@ def normalize(signal, reference_method='max', domain='auto',
+    the spectrum of a time-reversed and complex conjugated time signal, i.e.,
+    ``X_minus = fft(conj(x[::-1]))`` where the conjugate is required if the
+    input has complex-valued time data.
+
+    Parameters
+    ----------
+    signal_1 : Signal
+        The first input signal. It must have the same sample rate as
+        `signal_2` and its ``cshape`` must be broadcastable to that of
+        `signal_2`.
+    signal_2 : Signal
+        The second input signal. It must have the same sample rate as
+        `signal_1` and its ``cshape`` must be broadcastable to that of
+        `signal_1`.
+    mode : str, optional
+
+        ``'linear'``
+            Computes the linear correlation by zero padding the input signals
+            to a length of ``signal_1.n_samples + signal_2.n_samples - 1``
+            before applying the Fourier transform (see equations above).
+        ``'cyclic'``
+            Computes the cyclic correlation, which uses the input signals as
+            they are. In this case `signal_1` and `signal_2` must have the
+            same number of samples (length).
+
+        The default is ``'linear'``.
+
+    Returns
+    -------
+    correlation : numpy array
+        The correlation :math:`c[l]`. The shape of `correlation` matches the
+        shape to which `signal_1` and `signal_2` were broadcasted. The last
+        dimension of the shape is given by the `lags` (see below).
+    lags : numpy array
+        The time lags, i.e., the delays applied to `signal_2` (see equations
+        above). In case of a linear correlation, the time lags are in the
+        interval ``[-signal_2.n_samples + 1, signal_1.n_samples - 1]``. In case
+        of the cyclic correlation, they are in the interval
+        ``[-(signal_1.n_samples // 2) + 1, signal_1.n_samples // 2]`` if the
+        signals have an even number of samples, and in the interval
+        ``[-(signal_1.n_samples // 2), signal_1.n_samples // 2]`` if the
+        signals have an odd number of samples.
+    argmax : numpy array
+        The lag applied to `signal_2` which maximize the correlation between
+        the signals, i.e., :math:`\arg \max_{l} \, c[l]` for real-valued time
+        data, and :math:`\arg \max_{l} \, |c[l]|` for complex-valued time data.
+        The shape of `argmax` corresponds to the `cshape` to which the input
+        signals were broadcaseted.
+
+    Examples
+    --------
+
+    Linear and cyclic auto-correlation of a perfect sequence. Perfect sequences
+    have unit auto-correlation for :math:`l=0` and zero auto-correlation
+    otherwise
+
+    .. plot::
+
+        >>> import pyfar as pf
+        >>>
+        >>> signal = pf.signals.linear_perfect_sweep(2**8)
+        >>>
+        >>> for mode in ['linear', 'cyclic']:
+        >>>     cor, lags, _ = pf.dsp.correlate(signal, signal, mode)
+        >>>     result = pf.TimeData(cor / pf.dsp.energy(signal), lags)
```

Comment:
Why is this not part of the function?

Reply:
Do you mean the normlalization? I followed the scipy correlate definition that also does not do this.

Reply:
Sorry, I meant creating a time data object.